### PR TITLE
Add nullability inference by examining NRT annotations

### PR DESCRIPTION
**Pull request by Shane32**
Part 1 of #2837 -- add the ability to interpret nullability reference type annotations from user code.

### CODE REVIEW

**Comment by sungam3r** [comment]
Isn't there any ready-made package that provides nullable information?

**Comment by Shane32** [comment]
I don't know.  Even if there were, this code would eliminate a dependency.  Coverage should be at 100%, and I've covered every scenario I can think of.  As we can see from the api tests, even third party code is not always dependable.

**Comment by Shane32** [comment]
The actual executable code here is only 250 lines of code, and that includes about 80 lines of comments.

I've been using this code for quite a while in `Shane32.GraphQL.DI`, by the way.  It works very well from my experience.

**Comment by Shane32** [comment]
Of course I could package the code into a separate nuget package...

**Comment by Shane32** [comment]
Well coverage is about 100% except for the exceptions thrown if the NRT annotations are unable to be parsed... which makes it only 87.5%.  I forgot that I couldn't simulate those errors.

**Comment by sungam3r** [comment]
> Of course I could package the code into a separate nuget package...

@Shane32 I really advise you to pack this code into **source only** nuget package. I use them from time to time when I want to reuse some technical small code and do not want to have an external dependency in the created package.

File: src/GraphQL/Reflection/Nullability.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    /// </summary>
+    public enum Nullability
+    {
+        /// <summary>
+        /// The member is a reference type not marked with nullable reference annotations.
+        /// </summary>
+        Unknown = 0,
+
+        /// <summary>
+        /// The member is a non-nullable value type or is a nullable refernce type marked as non-nullable.
```
Comment (sungam3r) [inline src/GraphQL/Reflection/Nullability.cs:14]:
```suggestion
        /// The member is a non-nullable value type or is a nullable reference type marked as non-nullable.
```

File: src/GraphQL/Reflection/NullabilityInterpretationException.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Reflection;
+
+namespace GraphQL.Reflection
+{
+    /// <summary>
+    /// Exception when nullable reference annotations cannot be read from a parameter.
```
Comment (sungam3r) [inline src/GraphQL/Reflection/NullabilityInterpretationException.cs:7]:
```suggestion
    /// Exception when nullable reference annotations cannot be read from a member.
```

**Comment by Shane32** [comment]
> source only nuget package

Makes sense but I have not heard of that for C#.  Do you know of a repo with CI set up and so on that I can look at, and a repo that uses it?

File: src/GraphQL/Reflection/NullabilityInterpretationException.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    /// </summary>
+    public class NullabilityInterpretationException : Exception
+    {
+        /// <summary>
+        /// Initializes a new instance for the specified <paramref name="parameter"/>.
+        /// </summary>
+        public NullabilityInterpretationException(ParameterInfo parameter) : base($"Unable to interpret nullability attributes for argument '{parameter.Name}' on method '{parameter.Member.DeclaringType.Name}.{parameter.Member.Name}'.")
```
[the remark this answers is not in the document]
Comment (sungam3r) [reply src/GraphQL/Reflection/NullabilityInterpretationException.cs]: Why only one special ctor for parameter?

**Comment by sungam3r** [comment]
It will be easier for me to copy one of my projects and show.

File: src/GraphQL/Reflection/ReflectionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,199 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace GraphQL.Reflection
+{
+    /// <summary>
+    /// Provides extension methods to <see cref="ParameterInfo"/> and <see cref="MemberInfo"/> instances to read nullable type reference annotations.
```
Comment (sungam3r) [inline src/GraphQL/Reflection/ReflectionExtensions.cs:9]: PropertyInfo too. What about `MethodInfo`?

Comment (Shane32) [reply src/GraphQL/Reflection/ReflectionExtensions.cs:9]:
```suggestion
    /// Provides extension methods to <see cref="ParameterInfo"/> and <see cref="PropertyInfo"/> instances to read nullable type reference annotations.
```
All other methods are private

Comment (Shane32) [reply src/GraphQL/Reflection/ReflectionExtensions.cs:9]: Both arguments and the return value are parameters of a member.  So the ParameterInfo extension method covers those cases while the PropertyInfo extension method covers properties.

**Comment by sungam3r** [comment]
Sample csproj for building source-only package. Try it. For CI you may configure some props as we did it in Auth project to test different package versions.

```xml
<!--https://medium.com/@attilah/source-code-only-nuget-packages-8f34a8fb4738-->
<Project>
  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />

  <PropertyGroup>
    <NoBuild>true</NoBuild> <!-- Set false and comment Compile+CopyFilesToOutputDirectory targets to check build works -->
    <IsPackable>true</IsPackable>
    <TargetFramework>netstandard2.1</TargetFramework>
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild> <!-- NETSDK1085 The 'NoBuild' property was set to true but the 'Build' target was invoked. -->
    <Description>YOUR TEXT</Description>
    <NoWarn>$(NoWarn);NU5128</NoWarn>
  </PropertyGroup>

  <PropertyGroup Condition="'$(NoBuild)' == 'True'">
    <IncludeBuildOutput>false</IncludeBuildOutput>
    <ContentTargetFolders>contentFiles</ContentTargetFolders>
    <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
    <GenerateTargetFrameworkAttribute>false</GenerateTargetFrameworkAttribute>
    <SonarQubeExclude>true</SonarQubeExclude>
  </PropertyGroup>

  <ItemGroup Condition="'$(NoBuild)' == 'True'">
    <Compile Update="@(Compile)">
      <Pack>true</Pack>
      <PackagePath>$(ContentTargetFolders)\cs\$(TargetFramework)\$(PackageId)\%(RecursiveDir)</PackagePath>
    </Compile>
  </ItemGroup>

  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />

  <!-- Comment 2 targets to check build works. Alas, Condition attribute does not work on Target. -->
  <Target Name="Compile" />
  <Target Name="CopyFilesToOutputDirectory" />

</Project>
```

**Comment by Shane32** [comment]
Thanks!

**Comment by sungam3r** [comment]
Usage:
```
<PackageReference Include="PackageName" Version="1.0.0" PrivateAssets="all" />
```

PrivateAssets should be _all_ to not have any dependencies in consuming package. So source only package allows you to distribute your source code without dependencies.

**Comment by sungam3r** [comment]
>  <SonarQubeExclude>true</SonarQubeExclude>

You don't need this.

**Comment by sungam3r** [comment]
Should I continue review here or wait to review another repo?

**Comment by Shane32** [comment]
here

File: src/GraphQL/Reflection/ReflectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        private static (Nullability, IList<CustomAttributeTypedArgument>?) GetMethodParameterNullability(ParameterInfo parameter)
+        {
+            var nullableContext = GetMethodDefaultNullability(parameter.Member);
+
+            return ExamineMemberAttributes(nullableContext, parameter.CustomAttributes)
+                ?? throw new NullabilityInterpretationException(parameter);
+        }
+
+        /// <summary>
+        /// Returns the default NRT annotation for the member, and a list of the NRT annotations for the type.
```
Comment (sungam3r) [inline src/GraphQL/Reflection/ReflectionExtensions.cs:71]: > for the type

What type? Type of member? Declaring type?

File: src/GraphQL/Reflection/ReflectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            var list = new List<(Type, Nullability)>();
+            var index = 0;
+            Consider(property.PropertyType, false, list, nullabilityBytes, nullableContext, ref index);
+            if (nullabilityBytes != null && nullabilityBytes.Count != index)
+                throw new NullabilityInterpretationException(property);
+            return list;
+        }
+
+        /// <summary>
+        /// Returns the default NRT annotation for the parameter, and a list of the NRT annotations for the type.
```
Comment (sungam3r) [inline src/GraphQL/Reflection/ReflectionExtensions.cs:60]: What type? Type of member? Declaring type?

File: src/GraphQL/Reflection/ReflectionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+{
+    /// <summary>
+    /// Provides extension methods to <see cref="ParameterInfo"/> and <see cref="PropertyInfo"/> instances to read nullable type reference annotations.
+    /// </summary>
+    public static class ReflectionExtensions
+    {
+        /// <summary>
+        /// Examines the parameter type and walks the type definitions, for each returning the nullability information for the type.
+        /// <para>
+        /// For instance, for Test&lt;string, int&gt;, first returns nullability for Test&lt;string, int&gt;, then string, then int.
```
Comment (sungam3r) [inline src/GraphQL/Reflection/ReflectionExtensions.cs:16]: It is not clear from this example, in what order you need to traverse the tree of types. You provides a link to nullable-metadata docs so whatever. One can open and read.

**Review by sungam3r — COMMENTED** [review]
My thoughts are:
1. I understand what this code does.
2. I did not read it thoroughly.
3. A long time ago (in 2019 maybe) I added NRT inference support in [PublicApiApprover](https://github.com/PublicApiGenerator/PublicApiGenerator) and **it was a nightmare**. I read all that documentation and debugged the types traversal.
4. I definitely do not want to merge (and then maintain) such code in the main GraphQL.NET repository or in any other org repo. Similarly with tests. This is not a GraphQL.NET task.
5.  I'm fine with some external (preferably source-only) package (your or somebody else) to consume from GraphQL.NET.
6. Did you see https://github.com/SimonCropp/NullabilityInfo ?

So my conslusion is that you should to transfer all this code to another repo or use some other ready-made package. Search suggestions:
- https://www.nuget.org/packages?q=nullable
- https://www.nuget.org/packages?q=nullability

I can see a few source-only packages in that list.

**Comment by Shane32** [comment]
I know my own code; I know how much time I spent on this and how many situations I carefully covered. Personally, I don’t want to debug someone else’s code in this case.

I can easily make this into a separate nuget package. I’d make it a binary package, but we could have it also packaged as source only. If we want to go this route we could work on it together. 

I can review SimonCropp’s code … but this code is already done and verified for exactly our needs, including extensive testing.

**Comment by Shane32** [comment]
I looked around and in .NET 6 there is a built in class to read nullability information. (I wrote this code last year before .NET 6.) We should probably use the built in classes for .NET 6 and a source-compatible library for .NET Standard 2.0+ such as SimonCropp’s library. Or adjust mine to use the standardized interface. It seems that SimonCropp’s library is already in a source-only nuget package.

I suggest we keep the classes and tests here for two reasons: double check the third party library, and because the other tests I have written all use these same classes. We may eliminate the VerifyTestClasses tests if you like (since these tests verify the build process in order that nullability tests are testing what it was designed to. But I’d leave them in.

**Comment by sungam3r** [comment]
> I suggest we keep the classes and tests here for two reasons: 

I would not. Why not just add nessesary tests in other repo?

> I looked around and in .NET 6 there is a built in class to read nullability information.

🤔 Perhaps we can copy-paste built-in bits or maybe someone already did netstandard compatible package. I saw source only packages for nullability attributes in aforementioned nuget links.

**Comment by sungam3r** [comment]
By the way, what is the name of built-in nullability class?

**Comment by Shane32** [comment]
https://docs.microsoft.com/en-us/dotnet/api/system.reflection.nullabilityinfocontext?view=net-6.0

**Comment by Shane32** [comment]
> I would not. Why not just add nessesary tests in other repo?

Be fine if it was ours. We don’t necessarily have control of others’ repos. This is delicate code. I want to be sure it will work properly. Besides, I still need the test classes and most of the test setup for the rest of the tests. 

> 🤔 Perhaps we can copy-paste built-in bits or maybe someone already did netstandard compatible package. I saw source only packages for nullability attributes in aforementioned nuget links.

Like I said, SimonCropp’s package is a source only nuget package and is source compatible with the .NET 6 library. I’m not sure if he copied the source or not.

I just am quite busy to start investigating this. And then adapting all my code to use it.

**Comment by Shane32** [comment]
I wrote code to utilize the .NET 6 nullability inference code.  I can't get it to pass my tests.

**Pull request closed without merging**